File: src/rnet/packets/HandshakePacket.js
const RNetPacket = require("./RNetPacket");
//const Buffer = require("buffer");

class HandshakePacket extends RNetPacket {
    constructor(controllerID, handshakeType) {
        super();
        this.targetControllerID = controllerID;
        this.messageType = 0x02;
        this._handshakeType = handshakeType;
    }

    getMessageBody() {
        const buffer = Buffer.alloc(1);
        buffer.writeUInt8(this._handshakeType, 0);
        return buffer;
    }
}

HandshakePacket.fromPacket = function(rNetPacket) {
    if (rNetPacket instanceof RNetPacket) {
        const handshakePacket = new HandshakePacket();
        rNetPacket.copyToPacket(handshakePacket);
        handshakePacket._handshakeType = rNetPacket.messageBody.readUInt8(0);
        return handshakePacket;
    }
    else {
        throw new TypeError("Cannot create HandshakePacket with anything other than RNetPacket");
    }
}

module.exports = HandshakePacket;

File: src/RNetPi.Core/RNet/HandshakePacket.cs
using System;
using System.IO;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Handshake packet (MessageType = 0x02)
/// </summary>
public class HandshakePacket : RNetPacket
{
    public byte HandshakeType { get; set; }

    public HandshakePacket() : this(0, 0)
    {
    }

    public HandshakePacket(byte controllerID, byte handshakeType)
    {
        MessageType = 0x02;
        TargetControllerID = controllerID;
        HandshakeType = handshakeType;
    }

    protected override byte[] GetMessageBody()
    {
        return new byte[] { HandshakeType };
    }

    /// <summary>
    /// Creates a HandshakePacket from an RNetPacket
    /// </summary>
    public static HandshakePacket FromPacket(RNetPacket rnetPacket)
    {
        if (rnetPacket.MessageType != 0x02)
        {
            throw new ArgumentException("Cannot create HandshakePacket from packet with MessageType != 0x02");
        }

        var handshakePacket = new HandshakePacket();
        rnetPacket.CopyToPacket(handshakePacket);

        if (rnetPacket.MessageBody != null && rnetPacket.MessageBody.Length > 0)
        {
            handshakePacket.HandshakeType = rnetPacket.MessageBody[0];
        }

        return handshakePacket;
    }
}